google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsFirstCallable.java

```diff
@@ -16,23 +16,63 @@
 package com.google.cloud.bigtable.data.v2.stub.readrows;
 
 import com.google.api.core.ApiFuture;
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
 import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.ServerStreamingCallable;
+import com.google.api.gax.rpc.StateCheckingResponseObserver;
+import com.google.api.gax.rpc.StreamController;
 import com.google.api.gax.rpc.UnaryCallable;
 import com.google.cloud.bigtable.data.v2.models.Query;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Enhancement for `readRowsCallable().first()` to gracefully limit the row count instead of
  * cancelling the RPC
  */
-class ReadRowsFirstCallable<RowT> extends UnaryCallable<Query, RowT> {
-  private final UnaryCallable<Query, RowT> inner;
+@InternalApi
+public class ReadRowsFirstCallable<RowT> extends UnaryCallable<Query, RowT> {
 
-  ReadRowsFirstCallable(UnaryCallable<Query, RowT> inner) {
+  private final ServerStreamingCallable<Query, RowT> inner;
+
+  public ReadRowsFirstCallable(ServerStreamingCallable<Query, RowT> inner) {
     this.inner = inner;
   }
 
   @Override
   public ApiFuture<RowT> futureCall(Query query, ApiCallContext context) {
-    return inner.futureCall(query.limit(1), context);
+    ReadRowsFirstResponseObserver<RowT> observer = new ReadRowsFirstResponseObserver<>();
+    this.inner.call(query.limit(1), observer, context);
+    return observer.getFuture();
+  }
+
+  private class ReadRowsFirstResponseObserver<RowT> extends StateCheckingResponseObserver<RowT> {
+    private StreamController innerController;
+    private AtomicReference<RowT> newItem = new AtomicReference<>();
```

Comment:
I dont think this needs to be atomic, grpc should ensure a happens before relationship of onResponse & onComplete

Reply:
I think it's safer in case for some reason grpc still returned multiple rows? Then the response will still set to be the first row?